File: lib/shipr/hooks/github.rb
module Shipr
  module Hooks
    class GitHub < Grape::API
      logger Shipr.logger

      format :json
      default_format :json

      represent Shipr::Job, with: Entities::Job

      helpers do
        def event
          ActiveSupport::StringInquirer.new(headers['X-Github-Event'])
        end

        def deployments_service
          Shipr.deployments_service
        end
      end

      helpers do
        delegate :authenticate!, to: :warden

        def warden; env['warden'] end
      end

      before do
        authenticate!(scope: :api)
      end

      post do
        status 200
        if event.deployment?
          job = deployments_service.create(
            params.repository.full_name,
            sha:         params.deployment.sha,
            ref:         params.deployment.ref,
            guid:        params.deployment.id,
            force:       params.deployment.payload.try(:force),
            environment: params.deployment.environment,
            config:      params.deployment.payload.try(:config),
            description: params.deployment.description
          )
          present job
        else
          {}
        end
      end
    end
  end
end

File: server/api/github/github.go
package github

type Repository struct {
	FullName string `json:"full_name"`
}

type Deployment struct {
	ID          int64                  `json:"id"`
	Sha         string                 `json:"sha"`
	Ref         string                 `json:"ref"`
	Task        string                 `json:"task"`
	Environment string                 `json:"environment"`
	Payload     map[string]interface{} `json:"payload"`
	Description string
	Creator     struct {
		Login string `json:"login"`
	} `json:"creator"`
}

// DeploymentPayload is the webhook payload for a deployment event.
type DeploymentPayload struct {
	Deployment Deployment `json:"deployment"`
	Repository Repository `json:"repository"`
}
